lib/cdo/contact_rollups/v2/pardot_helpers.rb

```diff
@@ -38,33 +40,51 @@ def log(s)
 
   private
 
-  # Note: Pardot API key can become invalid and need to be refreshed midstream.
-  @@api_key = nil
-
-  # Authenticates and requests a new API key.
-  # API keys are valid for one hour while user keys are valid indefinitely.
-  # http://developer.pardot.com/#authentication
-  #
-  # @return [String] API key to use for subsequent requests
-  # @note This method has a side effect (it modifies a class variable) and may raise exception.
-  def request_api_key
-    doc = post_request(
-      AUTHENTICATION_URL,
-      {
-        email: CDO.pardot_username,
-        password: CDO.pardot_password,
-        user_key: CDO.pardot_user_key
-      }
+  PRIVATE_KEY = CDO.pardot_private_key
+  PARDOT_BUSINESS_ID = '0Uv5b0000004CHbCAM'
+
+  OAUTH_ENDPOINT = 'https://login.salesforce.com/services/oauth2/token'
+
+  @@access_token = nil
+
+  # Authenticates and requests an access token
+  # https://help.salesforce.com/articleView?id=sf.remoteaccess_oauth_jwt_flow.htm
+  # https://thespotforpardot.com/2021/02/02/pardot-api-and-getting-ready-with-salesforce-sso-users-part-3b-connecting-to-pardot-api-from-code/
+  def request_api_access_token
+    # build token payload
+    payload = {
+      # connected app client id
+      "iss": "3MVG9fMtCkV6eLhej.9tKBIE6OLmMCsxJAqIfy_eeSC1UaUR4rL0jkzUQOSRAhzfpHmUxUcuBp2JabX1ZOl2p",
+      # always login.salesforce.com
+      "aud": "https://login.salesforce.com",
+      # pardot account email
+      "sub": "plc-emails@code.org",
+      # less than 3 minutes after now
+      "exp": (Time.now + 2.minutes).to_i
```

Comment:
Probably a basic question, but would we just get new tokens repeatedly during a long-running job to send updates to Pardot? Is this similar to what happened previously?

Reply:
yep, each individual request to pardot will regenerate the token if it fails to auth, which is how it was set up before

Reply:
Ok cool, am I reading the old version correctly that we typically reauthorized after an hour? Just better practice to do it more frequently? Just asking for my understanding.

Reply:
oh yeah, its a requirement on salesforce's end for the exp field to be less than 3 min https://help.salesforce.com/s/articleView?id=sf.remoteaccess_oauth_jwt_flow.htm&type=5